File: antbook/p55/dptable.cpp
#include <iostream>

using namespace std;

int weight[] = {2,1,3,2};
int value[] = {3,2,4,2};
int weight_max = 5;
int itemnum = 4;

const int WEIGHT_MAX = 100;
const int ITEM_MAX = 100;
int dp[ITEM_MAX][WEIGHT_MAX];

void make_dptable()
{
    for( int i=0; i<WEIGHT_MAX; i++ ){
        dp[itemnum][i] = 0;
    }

    for( int i=itemnum-1; i>=0; i-- ){
        for( int j=0; j<WEIGHT_MAX; j++ ){
            if( j<weight[i] ){
                dp[i][j] = dp[i+1][j];
            }else{
                dp[i][j] = max( dp[i+1][j], dp[i+1][j-weight[i]]+value[i] );
            }
        }
    }
}

int main()
{
    make_dptable();
    cout << dp[0][weight_max] << endl;
}

File: antbook/p55/dptable.cs
using System;
using System.Text;
using System.Collections;
using System.Collections.Generic;
using System.Linq;

class dptable
{
    static int ITEM_MAX = 2000;
    static int WEIGHT_MAX = 2000;
    static int[,] table = new int[ITEM_MAX+1,WEIGHT_MAX+1];
    static int[] weight = {2,1,3,2};
    static int[] value = {3,2,4,2};
    static int itemnum = 4;
    
    public static void make_dptable(){
        for( int i=0; i<=WEIGHT_MAX; i++ ){
            table[itemnum,i] = 0;
        }

        for( int i=itemnum-1; i>=0; i-- ){
            for( int j=0; j<=WEIGHT_MAX; j++ ){
                if( j<weight[i] ){
                    table[i,j] = table[i+1,j];
                }else{
                    int[] a = {table[i+1,j], table[i+1,j-weight[i]]+value[i] };
                    table[i,j] = a.Max();
                }
            }
        }
        
    }

    public static void Main(){
        make_dptable();
//        int input[] = Console.ReadLine("hoge");
        Console.WriteLine(table[0,5]);
    }
}
